working to get xml and html with full origin tracking, for use in typepal

jurgenvinju opened the pull request.

jurgenvinju commented:
@DavyLandman this PR introduces origin tracking for XML, JSON, and HTML, based on what external parsers have to offer us in this regard (gson and jsoup). The goal is to use these parsers in IDEs for DSLs that have JSON or XML or HTML syntax.

Note that some of this information is designed for parse error reporting and not for origin tracking; so this could be brittle. If we end up in a future where jsoup and gson break the current contract, the test will fail and we'll have to write a grammar in Rascal instead. That is doable, but I'd rather reuse these libraries for now.

DavyLandman commented on src/org/rascalmpl/library/lang/html/AST.rsc:
```diff
@@ -9,8 +9,9 @@ This is HTML encoded like so:
+* <6> the `location` field is reserved for source locations of tags and other content. Normally this would be called `src` but `src` is reserved for other attributes in HTML
 }
-data HTMLElement // <1>
+data HTMLElement(loc location=|unknown:///|) // <1>
```
Shouldn't we use the same name as for parsetrees?

jurgenvinju replied:
I tried that, but that clashes with the `src` attribute of 5 standard html elements.

DavyLandman commented on src/org/rascalmpl/library/lang/html/IO.java:
```diff
@@ -113,23 +150,59 @@ else if (node instanceof DataNode) {
         }
+
+        return includeEndTags && e.isTracked()
+            ? factory.sourceLocation(file,
```
there appear to be no difference between these two branches?

jurgenvinju replied:
the variable names are too short too notice the difference. I'll refactor this.

DavyLandman commented on src/org/rascalmpl/library/lang/html/IO.rsc:
```diff
@@ -29,15 +29,15 @@ data HTMLSyntax
 }
 @javaClass{org.rascalmpl.library.lang.html.IO}
-java HTMLElement readHTMLString(str content, loc base=|http://localhost|);
+java HTMLElement readHTMLString(str content, loc base=|http://localhost|, bool trackOrigins=false, bool includeEndTags=true, loc src=|unknown:///|);
```
what is the differeence between `base` and `src`?

jurgenvinju replied:
base is the HTML base URI for the content and src is the actual file that this HTML is in. `Base` is actually used by the jsoup parser for implementing features in HTML, while src is only used for the origin tracking.

DavyLandman commented on src/org/rascalmpl/library/lang/json/internal/JsonValueReader.java:
```diff
@@ -52,20 +55,39 @@ public class JsonValueReader {
+        this.lineStartHandler = privateLookup.findVarHandle(JsonReader.class, "lineStart", int.class);
+      }
+      catch (NoSuchFieldException | SecurityException | IllegalAccessException e) {
+        // we disable the origin tracking if we can not get to the fields
```
nice fallback. maybe print and error message for us to detect that this happened?

DavyLandman commented on src/org/rascalmpl/library/lang/json/internal/JsonValueReader.java:
```diff
@@ -492,8 +574,16 @@ public IValue visitNode(Type type) throws IOException {
 
-        return vf.node("object", new IValue[] { }, kws);
+        if (src != null) {
+          return vf.node("object", new IValue[] { }, kws).asWithKeywordParameters().setParameter("src", vf.sourceLocation(src, startPos, endPos - startPos + 1, startLine, endLine, startCol, endCol + 1));
```
here the parameter is named `src`? not `location`?

jurgenvinju replied:
Yes because JSON is not HTML, so there is no a priori name clash there. However, it would be better if we test for a clash dynamically.

DavyLandman commented on src/org/rascalmpl/library/lang/rascal/tests/library/lang/html/IOTests.rsc:
```diff
@@ -12,3 +13,22 @@ test bool partialDocumentsAreCompletedOnRead()
 test bool partialDocumentsAreCompletedOnWrite() 
     = /html/ := writeHTMLString(p([text("paragraaf")]));
+
+test bool originTrackingElements() {
```
good for adding tests!

could we maybe do a small xml file, and read back the source location, and check that it indeed points to the right place?

DavyLandman commented on src/org/rascalmpl/uri/file/FileURIResolver.java:
```diff
@@ -192,11 +191,12 @@ public static ISourceLocation constructFileURI(String path) {
-		return null;
-	}
+
+        // Commented out during PR draft; this makes no sense. Is a test dependent on this?
```
This should not be commented out.

Some schemes allow us to detect charset, file does not.

jurgenvinju replied:
This is still puzzling me. The default method returns UTF8. should that not happen here too? Why are we returning null here?

jurgenvinju replied:
I'll revert it here. It needs a code comment at the very least.

DavyLandman reviewed (changes requested):
In general looks good, some small notes.

jurgenvinju commented on .vscode/launch.json:
```diff
@@ -30,7 +30,8 @@
             "mainClass": "org.rascalmpl.shell.RascalShell",
             "projectName": "rascal",
-            "vmArgs": "-Xss80m -Xmx2g"
+            "cwd": "/Users/jurgenv/git/rascal",
```
should launch.json be in the repo anyway? or can we remove it?

DavyLandman replied:
I think we should keep it in, just that we shouldn't put explicit paths in it. There are however vscode specific variables that you could use.

jurgenvinju commented:
* added tests for start line number positions for the XML and JSON parsers
* added `dropOrigin` feature for the XML, HTML and JSON writers such that the generated output does not contain the additional origin locations as attributes
* reverted the getCharSet method in FileURIResolver to the way it is supposed to be.
* Using "namespace" feature now to avoid collisions with the "src" attribute. If another attribute is called "src" then we use "rascal:src" or "rascal-src" depending on whether you are in the XML world or in Rascal.

I think this completes this branch.

DavyLandman commented:
>  such that the XML IO library could reuse the functionality of detecting the charset. 

this was actually abstracted away, as soon as you get a reader, we would either ask it from the scheme, and if they couldn't do it, use detect, without opening the stream twice. Is there no way to give the XML library a Reader instead of a InputStream?

DavyLandman commented on src/org/rascalmpl/uri/URIResolverRegistry.java:
```diff
@@ -882,7 +883,27 @@ public FileChannel getReadableFileChannel(ISourceLocation uri) throws IOExceptio
 		return resolver.getReadableFileChannel(uri);
 	}
 
+	public Charset detectCharset(ISourceLocation sloc) {
```
I'm not sure we should expose this. Normally you would call for a `getReader()` and just get it with the proper charset.

But it would avoid calling `getInputStream` twice.

jurgenvinju replied:
> this was actually abstracted away, as soon as you get a reader, we would either ask it from the scheme, and if they couldn't do it, use detect, without opening the stream twice. Is there no way to give the XML library a Reader instead of a InputStream?

I copied the method from one class to the other. There was no abstraction. This is about the writer, not the reader. Maybe that is a source of confusion?

DavyLandman replied:
how can you detect the charset of the writer if the target file doesn't exist yet? I don't think it makes sense to detect it at write time. For writing it should be an argument that's passed in (most likely with a default of utf8). 

Even for reading I think it would be better to not have this detection by default anymore, but that would break existing code.

jurgenvinju replied:
`writeFile` used this private method to find out if `inferCharSet` was true, what the charset should be from the previous contents. Especially in case of appending this is important. Now with writing XML files there are similar options to the writing methods (inferCharset and charset), and so we need the exact same functionality that was previously private for Prelude.java, now in lang/xml/IO.java

DavyLandman replied:
aah yes, I forgot about appending in the middle, do we support that? can you safely append an xml file?

jurgenvinju replied:
because our writers support offsets and lengths, we can always write into the middle of a file. Given the proper origins that we provide now, it becomes doable to safely replace a node inside an existing file.

jurgenvinju replied:
> how can you detect the charset of the writer if the target file doesn't exist yet? I don't think it makes sense to detect > it at write time. For writing it should be an argument that's passed in (most likely with a default of utf8).

We designed this together for `writeFile`; people can provide `charset`, but if they don't and they have `inferCharset` to true, we try to learn it from the existing file. This way we don't magically change the charset to utf8 of existing files just because the user of `writeFile` did not provide a charset explicitly. Also, it is absolutely necessary for append and write-into-middle as you said.

DavyLandman replied:
true, I forgot about appending files. This would be the first place outside of `writeFile` where we actually support writes to offsets in the file. if you feel we need to add that for this library, then I understand this change. A pitty we cannot make that functionality more generic, although for a lot of writers, we can't really guarantee any accurate behavior.

We should have at least one test that does this writing in the middle, and verify it's still a correct xml. (and not some kind of nested xml header in there.

jurgenvinju replied:
and it's also about writing the file back the way you found it, in terms of encoding, without having to think about it.
